horovod/tensorflow/__init__.py

```diff
@@ -223,3 +234,79 @@ def get_slot_names(self, *args, **kwargs):
     def variables(self, *args, **kwargs):
         """Calls this same method on the underlying optimizer."""
         return self._optimizer.variables(*args, **kwargs)
+
+
+class _DistributedGradientTape(tf.GradientTape):
+    """An tape that wraps another tf.GradientTape, using an allreduce to
+    average gradient values before applying gradients to model weights.
+
+    Args:
+      gradtape:
+        GradientTape to use for computing gradients and applying updates.
+      persistent:
+        Boolean controlling whether a persistent gradient tape
+        is created. False by default, which means at most one call can
+        be made to the gradient() method on this object.
+      watch_accessed_variables:
+        Boolean controlling whether the tape will
+        automatically `watch` any (trainable) variables accessed while the tape
+        is active. Defaults to True meaning gradients can be requested from any
+        result computed in the tape derived from reading a trainable `Variable`.
+        If False users must explicitly `watch` any `Variable`s they want to
+        request gradients from.
+      device_dense:
+        Device to be used for dense tensors. Uses GPU by default
+        if Horovod was build with HOROVOD_GPU_ALLREDUCE.
+      device_sparse:
+        Device to be used for sparse tensors. Uses GPU by default
+        if Horovod was build with HOROVOD_GPU_ALLGATHER.
+      compression:
+        Compression algorithm used during allreduce to reduce the amount
+        of data sent during the each parameter update step.  Defaults to
+        not using compression.
+      sparse_as_dense:
+        Treat all sparse gradients as dense tensors.  This can help improve
+        performance and memory utilization if the original sparse gradient
+        has high density.  Defaults to false.
+    """
+
+    def __init__(self, persistent, watch_accessed_variables,
+                 tape, device_dense='', device_sparse='',
+                 compression=Compression.none, sparse_as_dense=False):
+        super(self.__class__, self).__init__(persistent, watch_accessed_variables)
+        self._tape = tape
+        self._persistent = persistent
+        self._watch_accessed_variables = watch_accessed_variables
+        self._name = "Distributed"
+        self._device_dense = device_dense
+        self._device_sparse = device_sparse
+        self._compression = compression
+        self._sparse_as_dense = sparse_as_dense
+
+    def gradient(self, target, sources, output_gradients=None):
+        gradients = super(self.__class__, self).gradient(target, sources, output_gradients)
+        if size() > 1:
+            averaged_gradients = []
```

Comment:
Can you refactor this per the changes to `DistributedOptimizer` in #704?  Basically, without using `tf.contrib.eager.defun`, this will be very slow.

Reply:
ok. i updated it